### ISSUE TITLE
Is Shadowsocks draining battery by keeping mobile radio active more than necessary?

### ISSUE THREAD

**Issue by Mygod** [shadowsocks/shadowsocks-android#893]
Please answer these questions before submitting your issue. Thanks!

### What version of Shadowsocks are you using?

3.1.4.

### What Android OS version and device are you using?

7.0, bacon. (same ROM as the previous issue)

### What did you do?

Use Shadowsocks with cellular data on.

### What did you expect to see?

I'm not really sure.

### What did you see instead?

In my battery stat, the sum of every other app's mobile radio active time is around 55 mins. however Shadowsocks's mobile radio active time is ~80 mins which is much higher. Is Shadowsocks draining battery by keeping mobile radio active more than necessary?

### What is your config in detail (with all sensitive info masked)?

Encrypt method chacha20, OTA on, IPv6 route on, per-app proxy on, UDP off, KCP off, TFO off, VPN mode.

**Comment by madeye** [comment]
I think it's caused by other background services, e.g. Google Service. Shadowsocks itself won't send any packets.

BTW, what's your screen-on time?

**Comment by Mygod** [comment]
Current stat.
Screen on: 26m 6s (44 mAh)
Shadowsocks mobile radio active: 1h 4m 36s (200 mAh)
Other apps mobile radio active: (excluding bypassed apps) 39m 45s + 36s

**Comment by madeye** [comment]
What's the timeout settings for your server?

Also, can we do some experiments that disable shadowsocks for days and get the active mobile time?

From my experience in US with AT&T, the same phone with all the same apps and similar usage gives almost the same battery life (One day, Oneplus 2).

**Comment by Mygod** [comment]
My server is still using the same version as the previous commit.

Maybe. Could you test it? I need to use my device... as a phone?

I doubt whether doing so will yield comparable results in China.

**Comment by madeye** [comment]
The timeout should be large enough to avoid frequent reconnecting of GCM, say 600 second. 

I will try to disable GCM to see if we can get a much smaller active time.

**Comment by madeye** [comment]
What about this? https://m.reddit.com/r/Android/comments/44qx19/the_mobile_radio_battery_drain_bug_needs_more/

It looks like an analytics issue of battery life statistics.

**Comment by madeye** [comment]
Also see 

https://code.google.com/p/android/issues/detail?id=190396

Almost the same behavior.

**Comment by Mygod** [comment]
Yep. Seems like it.

> Google doesn't care

**Comment by Mygod** [comment]
Perhaps track battery usage for KCP as well? Some report battery drain increases after enabling KCP.

**Comment by Mygod** [comment]
Stat for today:
Screen on 2h 44m 1s
Shadowsocks 2h 56m 36s
Other apps 40m 48s (Android System) + 31m 47s (Android OS) + 2s (Google Play service)
Bypassed apps 32m 42s + 15m 26s + 13m 21s
Cell standby 7m 25s

**Comment by madeye** [comment]
If KCP enabled, high battery usage is expected, considering the keep-alive packets every 60 seconds.

**Comment by madeye** [comment]
Let's try 63fc69f. It may fix a "frequent reconnecting" issue for telegram like apps.

**Comment by Mygod** [comment]
Running 5d7e883.

Screen on 2h 35m 41s
Shadowsocks 1h 0m 12s
Other apps: 9m 9s (CyanogenMod System) + 7m 7s (Android OS) + 0s (Google Play services) + 0s (Spotify)
Bypassed apps 4m 35s

I think it's also related to how system calculates stat for different apps since I believe Spotify has definitely used more than 0s mobile active time.

**Comment by Mygod** [comment]
Still same commit.

Screen on 4h 47m 16s
Shadowsocks 3h 39m 19s
Other apps: 50m 6s (Android System) + 52m 46s (Android OS) + 0s (Google Play services) + 0s (Uber)
Bypassed apps: 53m 37s + 9m 51s + 16m 22s

What do you think?

**Comment by madeye** [comment]
I still think it's some kind of bug of Android itself. Actually, 4 screen-on hours are quite good.

**Comment by Mygod** [comment]
So let's close it for now?

Unfortunately, 4 screen-on hours is not possible for my device. I recharged it 3 times (110%~120%) during the day (not to 90%+). Here's the stat.

Total battery capacity 3100 mAh
Computed power use 3412 mAh
Observed power use 5022 mAh

**Comment by madeye** [comment]
Yep, let's close it for now.

**Comment by cs17899219** [comment]
Hi guys, 
    having tested on my Phone, I can reproduce  the problem "draining battery by keeping mobile radio active", The battery consumption is double on 4G LTE than on Wifi which is totally unacceptable in daily using.
I found the reason is  Shadowsocks mobile radio active is keeping prevent system  turn to sleep mode. Maybe some bug in VpnService or Badvpn.
I engaged in this issue and try workout what happen. 

I had downloaded the source code but it is not build by Gradle, Can anybody give me some quick guide on building local development environment. 

thanks.
